solr/solrj-zookeeper/src/java/org/apache/solr/common/cloud/ZkStateReader.java

```diff
@@ -377,8 +377,72 @@ private StatefulCollectionWatch compute(
     }
   }
 
-  private static class StatefulCollectionWatch extends CollectionWatch<DocCollectionWatcher> {
+  private class StatefulCollectionWatch extends CollectionWatch<DocCollectionWatcher> {
     private DocCollection currentState;
+
+    Watcher persistentWatcher;
+
+    final String coll;
+    private final String collectionPath;
+
+    private StatefulCollectionWatch(String coll) {
+      this.coll = coll;
+      this.collectionPath = DocCollection.getCollectionPath(coll);
+    }
+
+    public void handleWatch(WatchedEvent event) {
+      if (!collectionWatches.watchedCollections().contains(coll)) {
+        removePersistentWatch();
+        return;
+      }
+      log.debug("an event happened for {}, event: {}", coll, event.toString());
+      DocCollection collectionState = getCollection(coll);
+      if (collectionPath.equals(event.getPath())) {
+        DocCollection newState = null;
+        try {
+          newState = fetchCollectionState(coll, null);
+        } catch (KeeperException.SessionExpiredException
+            | KeeperException.ConnectionLossException e) {
+          log.warn("ZooKeeper watch triggered, but Solr cannot talk to ZK: ", e);
+        } catch (KeeperException e) {
+          log.error("exception for collection: [{}]", coll, e);
+          throw new ZooKeeperException(ErrorCode.SERVER_ERROR, "A ZK error has occurred", e);
+        } catch (InterruptedException e) {
+          Thread.currentThread().interrupt();
+          log.error("exception for collection: [{}]", coll, e);
+        }
+        collectionWatches.updateDocCollection(coll, newState);
+        synchronized (getUpdateLock()) {
+          constructState(Collections.singleton(coll));
+        }
+      }
+      if (collectionState!= null && collectionState.isPerReplicaState()) {
+        String path = event.getPath();
+        if ((event.getType() == EventType.NodeCreated || event.getType() == EventType.NodeDeleted)
+            && path.length() > collectionState.getZNode().length()
+            && path.startsWith(path)) {
```

Comment:
This part of the if statement will always be true as is, should this be more like:
```suggestion
            && path.startsWith(collectionPath)) {
```